manager/orchestrator/taskreaper/task_reaper.go

```diff
@@ -142,10 +142,10 @@ func (tr *TaskReaper) Run(ctx context.Context) {
 				if t.Status.State >= api.TaskStateOrphaned && t.ServiceID == "" {
 					tr.cleanup = append(tr.cleanup, t.ID)
 				}
-				// add tasks that have progressed beyond COMPLETE and have desired state REMOVE. These
-				// tasks are associated with slots that were removed as part of a service scale down
-				// or service removal.
-				if t.DesiredState == api.TaskStateRemove && t.Status.State >= api.TaskStateCompleted {
+				// add tasks that are yet unassigned or have progressed beyond COMPLETE, with
+				// desired state REMOVE. These tasks are associated with slots that were removed
+				// as part of a service scale down or service removal.
+				if t.DesiredState == api.TaskStateRemove && (t.Status.State >= api.TaskStateCompleted || t.Status.State < api.TaskStateAssigned) {
```

Comment:
Also, do we expect to get task update events for tasks which are marked REMOVE before they are assigned to the agent ? If a task is marked REMOVE, does the orchestrator even assign it to an agent ? @nishanttotla

Reply:
Good point @anshulpundir. It would be fair to expect the orchestrator to not assign those tasks at all. Let me check this.

Reply:
@anshulpundir so it seems like the dispatcher never checks desired state, only current state to decide to send down tasks: https://github.com/docker/swarmkit/blob/master/manager/dispatcher/dispatcher.go#L819

Does it make sense to check desired state here as well? Why should the dispatcher send a task to an agent if its desired state is past `SHUTDOWN`?

cc @stevvooe @cyli

Reply:
From discussion IRL with @nishanttotla:  It looks like the controller on the agent is responsible for looking at the desired state and acting on it (either by shutting down or not running at all).  The dispatcher sends down all the tasks that are assigned to the agent, and any changes to those tasks.  So if a task changes from `DesiredState` = `RUNNING` to `DesiredState` = `SHUTDOWN` it needs to send that update to the agent so that on the agent side the task gets shut down.  If the dispatcher skips sending the update, the task will just keep running on the agent.